prov/shm/src/smr_rma.c

```diff
@@ -71,16 +72,34 @@ static ssize_t smr_rma_fast(struct smr_region *peer_smr, const struct iovec *iov
 		goto signal;
 	}
 
-	memcpy(cma_iovec, iov, sizeof(*iov) * iov_count);
+	memcpy(vma_iovec, iov, sizeof(*iov) * iov_count);
 	for (i = 0; i < rma_count; i++) {
 		rma_iovec[i].iov_base = (void *) rma_iov[i].addr;
 		rma_iovec[i].iov_len = rma_iov[i].len;
 	}
 
 	total_len = ofi_total_iov_len(iov, iov_count);
 
-	ret = smr_cma_loop(peer_smr->pid, cma_iovec, iov_count, rma_iovec,
+#if HAVE_XPMEM
+	struct smr_domain *domain;
+	struct xpmem_client *xpmem = &smr_peer_data(ep->region)[peer_id].xpmem;
+
+	domain = container_of(ep->util_ep.domain, struct smr_domain,
+			      util_domain);
+
+	if (ep->region->xpmem_cap_self == SMR_VMA_CAP_ON &&
+	    xpmem->cap == SMR_VMA_CAP_ON) {
+		ret = ofi_shm_p2p_copy(FI_SHM_P2P_XPMEM, domain->xpmem_cache, vma_iovec,
+				iov_count, rma_iovec, rma_count, total_len, peer_id,
+				op == ofi_op_write, xpmem);
+	} else {
+		ret = smr_cma_loop(peer_smr->pid, vma_iovec, iov_count, rma_iovec,
+				   rma_count, 0, total_len, op == ofi_op_write);
+	}
+#else
+	ret = smr_cma_loop(peer_smr->pid, vma_iovec, iov_count, rma_iovec,
 			   rma_count, 0, total_len, op == ofi_op_write);
+#endif /* HAVE_XPMEM */
```

Comment:
Same comment as above.  Note that the 'else' and '#else' paths are identical.  They should follow the same path.

Larger question is why ofi_shm_p2p_copy() isn't configured correctly to just make the correct call?